File: 2016/Day7/main.dart
import 'dart:io';

bool containsAbba(String s) {
  for (int i = 0; i < s.length - 3; ++i) {
    if (s[i] != s[i + 1] && s[i] == s[i + 3] && s[i + 1] == s[i + 2]) {
      return true;
    }
  }
  return false;
}

Set<String> collectAba(String s) {
  Set<String> aba = {};
  for (int i = 0; i < s.length - 2; ++i) {
    if (s[i] != s[i + 1] && s[i] == s[i + 2]) {
      aba.add(s.substring(i, i + 3));
    }
  }
  return aba;
}

bool containsBab(String s, Set<String> abas) {
  for (int i = 0; i < s.length - 2; ++i) {
    if (s[i] != s[i + 1] && s[i] == s[i + 2]) {
      final bab = '${s[i + 1]}${s[i]}${s[i + 1]}';
      if (abas.contains(bab)) {
        return true;
      }
    }
  }
  return false;
}

void splitSegments(String ip, List<String> supernet, List<String> hypernet) {
  final re = RegExp(r'(\[|\])');
  final matches = re.allMatches(ip);
  var insideBrackets = false;
  var lastIndex = 0;

  for (var match in matches) {
    if (match.start > lastIndex) {
      final segment = ip.substring(lastIndex, match.start);
      if (insideBrackets) {
        hypernet.add(segment);
      } else {
        supernet.add(segment);
      }
    }
    insideBrackets = match.group(0) == '[';
    lastIndex = match.end;
  }

  if (lastIndex < ip.length) {
    final segment = ip.substring(lastIndex);
    if (insideBrackets) {
      hypernet.add(segment);
    } else {
      supernet.add(segment);
    }
  }
}

void main() async {
  final ips = await File('input.txt').readAsLines();

  var tlsCount = 0;
  var sslCount = 0;

  for (var ip in ips) {
    final supernet = <String>[];
    final hypernet = <String>[];
    splitSegments(ip, supernet, hypernet);

    var hasAbbaOutside = false;
    var hasAbbaInside = false;

    for (var segment in supernet) {
      if (containsAbba(segment)) {
        hasAbbaOutside = true;
        break;
      }
    }

    for (var segment in hypernet) {
      if (containsAbba(segment)) {
        hasAbbaInside = true;
        break;
      }
    }

    if (hasAbbaOutside && !hasAbbaInside) {
      tlsCount++;
    }

    final abas = <String>{};
    for (var segment in supernet) {
      abas.addAll(collectAba(segment));
    }

    var supportsSsl = false;
    for (var segment in hypernet) {
      if (containsBab(segment, abas)) {
        supportsSsl = true;
        break;
      }
    }

    if (supportsSsl) {
      sslCount++;
    }
  }

  print('Part 1: $tlsCount');
  print('Part 2: $sslCount');
}

File: 2016/Day7/main.py
import re
import sys

def contains_abba(s: str) -> bool:
    for i in range(len(s) - 3):
        if s[i] != s[i + 1] and s[i] == s[i + 3] and s[i + 1] == s[i + 2]:
            return True
    return False

def collect_aba(s: str) -> set[str]:
    abas: set[str] = set()
    for i in range(len(s) - 2):
        if s[i] != s[i + 1] and s[i] == s[i + 2]:
            abas.add(s[i:i + 3])
    return abas

def contains_bab(s: str, abas: set[str]) -> bool:
    for i in range(len(s) - 2):
        if s[i] != s[i+1] and s[i] == s[i+2]:
            bab_to_aba = s[i+1] + s[i] + s[i+1]
            if bab_to_aba in abas:
                return True
    return False

def split_segments(ip: str) -> tuple[list[str], list[str]]:
    supernet: list[str] = []
    hypernet: list[str] = []

    hypernet_segments = re.findall(r'\[([a-z]+)\]', ip)
    hypernet.extend(hypernet_segments)

    placeholder = '@@@'
    modified_ip = re.sub(r'\[([a-z]+)\]', placeholder, ip)

    supernet_parts = modified_ip.split(placeholder)
    supernet.extend(filter(None, supernet_parts))

    return supernet, hypernet

if __name__ == '__main__':
    ips = [line.strip() for line in sys.stdin if line.strip()]

    tls_count = 0
    ssl_count = 0

    for ip in ips:
        supernet_segments, hypernet_segments = split_segments(ip)

        has_abba_outside = False
        for segment in supernet_segments:
            if contains_abba(segment):
                has_abba_outside = True
                break

        has_abba_inside = False
        for segment in hypernet_segments:
            if contains_abba(segment):
                has_abba_inside = True
                break

        if has_abba_outside and not has_abba_inside:
            tls_count += 1

        all_abas: set[str] = set()
        for segment in supernet_segments:
            all_abas.update(collect_aba(segment))

        supports_ssl = False
        if all_abas:
            for segment in hypernet_segments:
                if contains_bab(segment, all_abas):
                    supports_ssl = True
                    break

        if supports_ssl:
            ssl_count += 1

    print(f'Part 1: {tls_count}')
    print(f'Part 2: {ssl_count}')
